### PR TITLE
Add variable doc build version for github links

### PR DESCRIPTION
**Pull request by SYangster**
Use sphinx extlinks extension

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [x] Documentation updated.

### CODE REVIEW

File: docs/example_applications_algorithms.rst [hunk 1]
```diff
@@ -26,56 +26,56 @@ The following tutorials and quickstart guides walk you through some of these exa
   1. **Hello World** introduction to NVFlare.
 
      1.1. Deep Learning to Federated Learning
-          * `Deep Learning to Federated Learning (GitHub) <https://github.com/NVIDIA/NVFlare/blob/main/examples/hello-world/ml-to-fl>`_ - Example for converting Deep Learning (DL) to Federated Learning (FL).
+          * :github_nvflare_link:`Deep Learning to Federated Learning (GitHub) <examples/hello-world/ml-to-fl>` - Example for converting Deep Learning (DL) to Federated Learning (FL).
```
[the remark this answers is not in the document]
Comment (SYangster) [reply docs/example_applications_algorithms.rst:29]: No, this is a sphinx extension, hence it won't show up in github's rst rendering, but does work once the docs are built

File: docs/example_applications_algorithms.rst [hunk 2]
```diff
@@ -4,7 +4,7 @@
 Example Applications
 ####################
 NVIDIA FLARE has several tutorials and examples to help you get started with federated learning and to explore certain features in
-`the examples directory <https://github.com/NVIDIA/NVFlare/tree/main/examples>`_.
+:github_nvflare_link:`the examples directory <examples>`.
```
[the remark this answers is not in the document]
Comment (SYangster) [reply docs/example_applications_algorithms.rst:7]: :github_nvflare_link: will convert to use the base url defined in conf.py

https://www.sphinx-doc.org/en/master/usage/extensions/extlinks.html

File: docs/key_features.rst [hunk 3]
```diff
@@ -62,10 +62,10 @@ Learning Algorithms
       a set of initial weights is distributed to client Workers who perform local training.  After local training,
       clients return their local weights as a Shareables that are aggregated (averaged).  This new set of global average
       weights is redistributed to clients and the process repeats for the specified number of rounds.
-    - `FedProx <https://arxiv.org/abs/1812.06127>`_ (example configuration can be found in cifar10_fedprox of `CIFAR-10 example <https://github.com/NVIDIA/NVFlare/tree/main/examples/cifar10>`_) -
+    - :github_nvflare_link:`FedProx <examples/cifar10>`) -
```
Comment (YuanTingHsieh) [inline docs/key_features.rst:65]: this link is wrong, should be <examples/advanced/cifar10>

File: docs/key_features.rst [hunk 4]
```diff
@@ -62,10 +62,10 @@ Learning Algorithms
...
       implements a :class:`Loss function <nvflare.app_common.pt.pt_fedproxloss.PTFedProxLoss>` to penalize a client's
       local weights based on deviation from the global model.
-    - `FedOpt <https://arxiv.org/abs/2003.00295>`_ (example configuration can be found in cifar10_fedopt of `CIFAR-10 example <https://github.com/NVIDIA/NVFlare/tree/main/examples/cifar10>`_) -
+    - :github_nvflare_link:`FedOpt <examples/cifar10>`) -
```
Comment (YuanTingHsieh) [inline docs/key_features.rst:68]: this link is wrong, should be <examples/advanced/cifar10>

File: docs/programming_guide/controllers/cross_site_model_evaluation.rst [hunk 5]
```diff
@@ -11,19 +11,19 @@ client dataset.
...
+The :github_nvflare_link:`hello-numpy-cross-val example <examples/hello-world/hello-numpy-cross-val>` is a simple
 example that implements the :class:`cross site model evaluation workflow<nvflare.app_common.workflows.cross_site_model_eval.CrossSiteModelEval>`.
 
 .. note::
 
    Previously in NVFlare before version 2.0, cross-site validation was built into the framework itself, and there was an
    admin command to retrieve cross-site validation results. In NVFlare 2.0, with the ability to have customized
    workflows, cross-site validation is no longer in the NVFlare framework but is instead handled by the workflow. The
-   the `cifar10 example <https://github.com/NVIDIA/NVFlare/tree/main/examples/cifar10>`_ is configured to run cross-site
+   the :github_nvflare_link:`cifar10 example <examples/cifar10>` is configured to run cross-site
```
Comment (YuanTingHsieh) [inline docs/programming_guide/controllers/cross_site_model_evaluation.rst:22]: this link is wrong, should be <examples/advanced/cifar10>

File: docs/programming_guide/controllers/cyclic_workflow.rst [hunk 6]
```diff
@@ -12,5 +12,5 @@ flow implemented through the :meth:`control_flow()<control_flow>` method based o
 
 Example with Cyclic Workflow
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-See the `Hello Cyclic Example <https://github.com/NVIDIA/NVFlare/tree/main/examples/hello-cyclic>`_ for an example application with
+See the :github_nvflare_link:`Hello Cyclic Example <examples/hello-cyclic>` for an example application with
```
Comment (YuanTingHsieh) [inline docs/programming_guide/controllers/cyclic_workflow.rst:15]: this link is wrong, should be <examples/hello-world/hello-cyclic>

File: docs/programming_guide/controllers/scatter_and_gather_workflow.rst [hunk 7]
```diff
@@ -45,5 +45,5 @@ Below is the signature for an aggregator.
 
 Example with Scatter and Gather Workflow
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-See the `Hello Numpy Scatter and Gather Example <https://github.com/NVIDIA/NVFlare/tree/main/examples/hello-numpy-sag>`_ for an example application with
+See the :github_nvflare_link:`Hello Numpy Scatter and Gather Example <examples/hello-numpy-sag>` for an example application with
```
Comment (YuanTingHsieh) [inline docs/programming_guide/controllers/scatter_and_gather_workflow.rst:48]: this link is wrong, should be <examples/hello-world/hello-numpy-sag>

File: docs/programming_guide/filters.rst [hunk 8]
```diff
@@ -44,7 +44,7 @@ purpose of security. In fact, privacy and homomorphic encryption techniques are
     - SVTPrivacy for differential privacy through sparse vector techniques (:mod:`nvflare.app_common.filters.svt_privacy`)
     - Homomorphic encryption filters to encrypt data before sharing (:mod:`nvflare.app_common.homomorphic_encryption.he_model_encryptor.py` and :mod:`nvflare.app_common.homomorphic_encryption.he_model_decryptor`)
 
-For an example application using SVTPrivacy, see `Differential Privacy for BraTS18 segmentation (GitHub) <https://github.com/NVIDIA/NVFlare/tree/main/examples/brats18>`_.
+For an example application using SVTPrivacy, see :github_nvflare_link:`Differential Privacy for BraTS18 segmentation (GitHub) <examples/brats18>`.
```
Comment (YuanTingHsieh) [inline docs/programming_guide/filters.rst:47]: this link is wrong

File: docs/real_world_fl/FLAdminAPI.rst [hunk 9]
```diff
@@ -34,7 +34,7 @@ The ``logout()`` in AdminAPI handles ending the overseer_agent which may take a
 
 Usage
 -----
-See the example scripts ``run_fl.py`` in `CIFAR-10 <https://github.com/NVIDIA/NVFlare/tree/main/examples/cifar10>`_ for
+See the example scripts ``run_fl.py`` in :github_nvflare_link:`CIFAR-10 <examples/cifar10>` for
```
Comment (YuanTingHsieh) [inline docs/real_world_fl/FLAdminAPI.rst:37]: this link is wrong

File: docs/conf.py [hunk 10]
```diff
@@ -92,6 +95,7 @@ def resolve_xref(self, env, fromdocname, builder, typ, target, node, contnode):
 # This pattern also affects html_static_path and html_extra_path.
 exclude_patterns = []
 
+extlinks = {"github_nvflare_link": (f"https://github.com/NVIDIA/NVFlare/tree/{build_version}/%s", "")}
```
Comment (YuanTingHsieh) [inline docs/conf.py:98]: NVFlare/tree or NVFlare/blob?

**Review by YuanTingHsieh — COMMENTED** [review]
Some missing links, otherwise good job.

File: docs/conf.py (hunk 10 above)
Comment (SYangster) [reply docs/conf.py:98]: Currently github redirects to either tree or blob based on if it is a directory or a file, so the link will work regardless. I put one version using tree for simplicity, but should we instead have two versions for blob/tree for clarity?

Comment (YuanTingHsieh) [reply docs/conf.py:98]: Got it, thanks for the explanation, either way is good to me

**Review by YuanTingHsieh — APPROVED** [review]

**Pull request merged**